core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala

```diff
@@ -0,0 +1,212 @@
+
+    if (shuffleService == null) {
+      sparkConf.set("spark.shuffle.service.enabled", "true")
+      shuffleService = new StandaloneWorkerShuffleService(sparkConf, new SecurityManager(sparkConf))
+      shuffleService.startIfEnabled()
+    }
+  }
+
+  override def launchTask(d: ExecutorDriver, taskInfo: TaskInfo) {
+    if (executorProc != null) {
+      logError("Received LaunchTask while executor is already running")
+      val status = TaskStatus.newBuilder()
+        .setTaskId(taskInfo.getTaskId)
+        .setSlaveId(taskInfo.getSlaveId)
+        .setState(TaskState.TASK_FAILED)
+        .setMessage("Received LaunchTask while executor is already running")
+        .build()
+      d.sendStatusUpdate(status)
+      return
+    }
+
+    killed = false
+
+    logInfo("Launching task id: " + taskInfo.getTaskId.getValue)
+
+    // We are launching the CoarseGrainedExecutorBackend via subprocess
+    // because the backend is designed to run in its own process.
+    // Since it's a shared class we are preserving the existing behavior
+    // and launching it as a subprocess here.
+    val command = "exec " + Utils.deserialize[String](taskInfo.getData().toByteArray)
+
+    logInfo("Running command: " + command)
+
+    // Mesos only work on linux platforms, so we assume bash is available is Mesos is used.
+    val pb = new ProcessBuilder("/bin/bash", "-c", command)
+
+    val currentEnvVars = pb.environment()
+    for (variable <- taskInfo.getExecutor.getCommand.getEnvironment.getVariablesList()) {
+      currentEnvVars.put(variable.getName, variable.getValue)
+    }
+
+    executorProc = pb.start()
+
+    new Thread("stderr reader for task " + taskInfo.getTaskId.getValue) {
+      override def run() {
+        for (line <- Source.fromInputStream(executorProc.getErrorStream).getLines) {
+          System.err.println(line)
+        }
+      }
+    }.start()
+
+    new Thread("stdout reader for task " + taskInfo.getTaskId.getValue) {
+      override def run() {
+        for (line <- Source.fromInputStream(executorProc.getInputStream).getLines) {
+          System.out.println(line)
+        }
+      }
+    }.start()
+
+    driver.sendStatusUpdate(TaskStatus.newBuilder()
+      .setState(TaskState.TASK_RUNNING)
+      .setTaskId(taskInfo.getTaskId)
+      .build)
+
+    new Thread("process waiter for mesos executor for task " + taskInfo.getTaskId.getValue) {
+      override def run() {
+        executorProc.waitFor()
+        val (state, msg) = if (killed) {
+          (TaskState.TASK_KILLED, "")
+        }  else if (executorProc.exitValue() == 0) {
+          (TaskState.TASK_FINISHED, "")
+        } else {
+          (TaskState.TASK_FAILED, "Exited with status: " + executorProc.exitValue().toString)
+        }
+        // We leave the shuffle service running after the task.
+        cleanup(state, msg)
+      }
+    }.start()
+
+    taskId = taskInfo.getTaskId
+  }
+
+  override def error(d: ExecutorDriver, message: String) {
```

Comment:
methods without return types should use `: Unit` (see https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide)